File: perf/benchmarks/micro/gcd_loop/gcd_loop.cpp
// gcd_loop — Euclidean GCD tight-loop micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

static int64_t gcd(int64_t a, int64_t b) {
    while (b != 0) {
        int64_t t = a % b;
        a = b;
        b = t;
    }
    return a;
}

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 2000000;
    auto t0 = std::chrono::steady_clock::now();
    int64_t cs = 0;
    int64_t seed = 1;
    for (int64_t i = 1; i <= n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        cs = (cs + gcd(i, seed)) & 0x7FFFFFFFLL;
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/gcd_loop/gcd_loop.py
# gcd_loop — Euclidean GCD tight-loop micro-benchmark.
#
# For i in [1, N] compute gcd(i, b_i) where b_i comes from a deterministic
# LCG, and accumulate the sum modulo 2^31. Argument: N (default 2_000_000).

import sys
import time


def gcd(a: int, b: int) -> int:
    while b != 0:
        a, b = b, a % b
    return a


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 2_000_000
    t0 = time.perf_counter_ns()
    cs = 0
    seed = 1
    for i in range(1, n + 1):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        cs = (cs + gcd(i, seed)) & 0x7FFFFFFF
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
